### ISSUE TITLE
Error during class documentation phase - CompletionException: NoSuchElementException / IndexOutOfBoundsException

### ISSUE THREAD

**Issue by lievendf** [vscode-abl/vscode-abl#217]
```
[2024-09-13T09:22:48.224+0200] [ERROR] [<project name here>] Error during class documentation phase
java.util.concurrent.CompletionException: java.util.NoSuchElementException
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.thenApply(Unknown Source)
	at eu.rssw.openedge.ls.builder.x.run(SourceFile:171)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.util.NoSuchElementException
	at java.base/java.util.LinkedList.getLast(Unknown Source)
	at org.prorefactor.proparse.PostLexer.preproElse(PostLexer.java:181)
	at org.prorefactor.proparse.PostLexer.nextToken(PostLexer.java:90)
	at org.prorefactor.proparse.ProparseSkipFilter.nextToken(ProparseSkipFilter.java:46)
	at org.prorefactor.proparse.NameDotTokenFilter.fillHeap(NameDotTokenFilter.java:54)
	at org.prorefactor.proparse.NameDotTokenFilter.nextToken(NameDotTokenFilter.java:152)
	at org.prorefactor.proparse.TokenList.fillHeap(TokenList.java:48)
	at org.prorefactor.proparse.TokenList.nextToken(TokenList.java:139)
	at org.prorefactor.proparse.FunctionKeywordTokenFilter.nextToken(FunctionKeywordTokenFilter.java:57)
	at org.prorefactor.proparse.ABLLexer.nextToken(ABLLexer.java:135)
	at org.antlr.v4.runtime.BufferedTokenStream.fetch(BufferedTokenStream.java:169)
	at org.antlr.v4.runtime.BufferedTokenStream.sync(BufferedTokenStream.java:152)
	at org.antlr.v4.runtime.BufferedTokenStream.consume(BufferedTokenStream.java:136)
	at org.antlr.v4.runtime.Parser.consume(Parser.java:571)
	at org.antlr.v4.runtime.Parser.match(Parser.java:205)
	at org.prorefactor.proparse.antlr4.Proparse.statementEnd(Proparse.java:63772)
	at org.prorefactor.proparse.antlr4.Proparse.bufferCopyStatement(Proparse.java:19115)
	at org.prorefactor.proparse.antlr4.Proparse.statement(Proparse.java:3444)
	at org.prorefactor.proparse.antlr4.Proparse.blockOrStatement(Proparse.java:1604)
	at org.prorefactor.proparse.antlr4.Proparse.codeBlock(Proparse.java:1505)
	at org.prorefactor.proparse.antlr4.Proparse.procedureStatement(Proparse.java:58077)
	at org.prorefactor.proparse.antlr4.Proparse.statement(Proparse.java:4165)
	at org.prorefactor.proparse.antlr4.Proparse.blockOrStatement(Proparse.java:1604)
	at org.prorefactor.proparse.antlr4.Proparse.program(Proparse.java:1440)
	at org.prorefactor.treeparser.ParseUnit.parse(ParseUnit.java:340)
	at org.prorefactor.treeparser.ParseUnit.parse(ParseUnit.java:309)
	at org.prorefactor.treeparser.ParseUnit.treeParser01(ParseUnit.java:402)
	at eu.rssw.openedge.ls.builder.y.a(SourceFile:39)
	at eu.rssw.openedge.ls.builder.y.apply(SourceFile:26)
	... 7 more
```
...
```
[2024-09-13T09:35:47.315+0200] [ERROR] [Glims] Error during class documentation phase
java.util.concurrent.CompletionException: java.lang.IndexOutOfBoundsException
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.thenApply(Unknown Source)
	at eu.rssw.openedge.ls.builder.x.run(SourceFile:157)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.IndexOutOfBoundsException
	at java.base/java.nio.ByteBuffer.wrap(Unknown Source)
	at eu.rssw.pct.RCodeInfo.processHeader(RCodeInfo.java:172)
	at eu.rssw.pct.RCodeInfo.<init>(RCodeInfo.java:118)
	at eu.rssw.pct.RCodeInfo.<init>(RCodeInfo.java:105)
	at eu.rssw.openedge.ls.builder.z.a(SourceFile:37)
	at eu.rssw.openedge.ls.builder.z.apply(SourceFile:26)
	... 7 more
```

Could these be attributed to multiple compilation sessions getting in each others way?

**Comment by gquerret** [comment]
Stack trace #1 : proparse error. Please open new issue if you can reproduce the problem. That can't be solved without the source code (it's a preprocessor issue).

Stack trace #2 : multiple threads reading / writing the same file in parallel, that won't be fixed for now. Stack trace might be hidden in the future.
